gemfirexd/core/src/main/java/com/pivotal/gemfirexd/internal/engine/distributed/message/LeadNodeExecutorMsg.java

```diff
@@ -128,41 +130,58 @@ protected void execute() throws Exception {
       SnappyResultHolder srh = new SnappyResultHolder(exec);
 
       srh.prepareSend(this);
-      //sendresult and lastResult is called via prepareSend
-//      this.sendResult(srh);
-//      this.lastResult(srh);
       this.lastResultSent = true;
       this.endMessage();
       if (GemFireXDUtils.TraceQuery) {
         SanityManager.DEBUG_PRINT(GfxdConstants.TRACE_QUERYDISTRIB,
                 "LeadNodeExecutorMsg.execute: Sent Last result ");
       }
     } catch (Exception ex) {
-      // Catch all exceptions and convert so can be caught at XD side
-      Throwable cause = ex;
-      while (cause != null) {
-        if (cause.getClass().getName().contains("AnalysisException")) {
-          throw StandardException.newException(
-              SQLState.LANG_UNEXPECTED_USER_EXCEPTION, cause, cause.getMessage());
-        } else if (cause.getClass().getName().contains("apache.spark.storage")) {
-          throw StandardException.newException(
-              SQLState.DATA_UNEXPECTED_EXCEPTION, cause, cause.getMessage());
-        } else if (cause.getClass().getName().contains("apache.spark.sql")) {
-          Throwable nestedCause = cause.getCause();
-          while (nestedCause != null) {
-            if (nestedCause.getClass().getName().contains("ErrorLimitExceededException")) {
-              throw StandardException.newException(
-                  SQLState.LANG_UNEXPECTED_USER_EXCEPTION, nestedCause, nestedCause.getMessage());
-            }
-            nestedCause = nestedCause.getCause();
+      throw getExceptionToSendToServer(ex);
+    }
+  }
+
+  private static class SparkExceptionWrapper extends Exception {
+    public SparkExceptionWrapper(Exception ex) {
+      super(ex.getClass().getName() + ": " + ex.getMessage(), ex.getCause());
+      this.setStackTrace(ex.getStackTrace());
+    }
+  }
+
+  private Exception getExceptionToSendToServer(Exception ex) {
+    // Catch all exceptions and convert so can be caught at XD side
+    // Check if the exception can be serialized or not
+    SparkExceptionWrapper wrapperEx = null;
```

Comment:
As said below in all the cases nested should go. The naming a little confusion. If you see catually threre is a loop and the nested exception are being looked into. Making the changes accordingly.